src/CachePeers.h

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_CACHEPEERS_H
+#define SQUID_CACHEPEERS_H
+
+#include "base/forward.h"
+#include "CachePeer.h"
+#include "mem/PoolingAllocator.h"
+
+#include <vector>
+
+/// Weak pointers to zero or more Config.peers.
+/// Users must specify the selection algorithm and the order of entries.
+using SelectedCachePeers = std::vector< CbcPointer<CachePeer>, PoolingAllocator< CbcPointer<CachePeer> > >;
+
+/// Temporary, local storage of raw pointers to zero or more Config.peers.
+using RawCachePeers = std::vector<CachePeer *, PoolingAllocator<CachePeer*> >;
```

Comment:
Please remove this type. We should not be storing raw-pointers for these in any new code. The temporary vector should be storing the relevant `Pointer` type instead.

Reply:
> We should not be storing raw-pointers for these in any new code.

I disagree with that assertion. Given the expense associated with smart pointers, especially cbdata pointers, _temporary_ storage of raw pointers is the best solution for use cases like this:

```C++
if (const auto raw = smart.valid())
    raw->doSomething();
```

This type is used for exactly the same purpose; the only difference is that the corresponding method call has to be split across two loops, necessitating (temporary) storage between the loops.

> The temporary vector should be storing the relevant `Pointer` type instead.

Why? The current PR code that uses this type for temporary (function execution-scoped) storage of selected/validated cache peers is better than the proposed code that would abuse a cbdata-protected pointer for the same purpose.

The only viable alternative that I can see is storing [references](https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper) instead of raw pointers. However, since we then need to create CbcPointers, I think the current approach is slightly better -- creating a long-term stored pointer from a reference raises arguably bigger red flags that temporary storing a raw pointer.

Reply:
Under the principles of C++ RAII the use of ```vector``` may result in either leaked memory allocation, or crashes on double-free. Assuming that all code changes in future will always correctly account for both those risks is a **Bad Practice**.

Relating to performance, the heavy memory operations from ```Pointer``` are largely offset against the removal of delayed conversion to ```Pointer```. The remaining weight is largely due to inefficiency in our cbdata ```Pointer``` design which is a separate performance issue which will gain Squid much larger benefits if fixed at the definition itself.